### PR TITLE
Skins :: adjust styles of skin context menus, tooltip… 2nd attempt

### PR DESCRIPTION
**Pull request by ronso0**
Synchronise styles of skin right-click menus, tooltip, ...
[2nd attempt after #2337 was merged too early. See conversation there...]
[hopefully all changes are included here. ]

This should definitely be included in the 2.3 release because it makes the entire regular GUI interaction a consistent experience. No more OS themes crushing in, except popup windows like file dialogs and warnings.

It affects
* Tracks table menu
* hotcue menu
* cover art menu
* tooltips
* effect selector drop-down
* table header menu
* editable fields (Search box, beat spinboxes, library items)
* skin settings menu (Tango, LateNight, maybe Deere)
* sidebar menu

This is how the popups etc. are supposed to look like:
menus-popups-etc_deere
menus-popups-etc_latenight
menus-popups-etc_shade-dark
menus-popups-etc_tango

ToDo
- [x] ~~use relative font size where applicable?~~ Nope, maybe later on..
- [x] style indeterminate (partially checked) checkboxes (tracks' Crates menu)
- [x] style editable fields
- [x] Shade: fix library selection colors in all color schemes
- [x] include AutoDJ trasition mode drop-down from #2103 
- [x] remove WEffectSelector style hack from MacOS stylesheets
- [x] restrict QScrollBar, QTreeView, QTableView and QTextBrowser styles to Mixxx widgets like the library, WEffectSelector and #fadeModeComboBox (don't style QFileDialogs etc.)
- [x] restrict QHeaderView styles to Mixxx widgets (#Library)
- [x] restrict QMenu styles to Mixxx widgets (#Library, WOverView, WSEarchLineEdit, QSpinBox)
- [x] include `WCueMenuPopup` #2362 

Please test with various themes, also on Windows and MacOS, also without desktop styles (window manager olnly, like i3wm)

### CODE REVIEW

**Comment by ronso0** [comment]
It would be nice if someone with MacOS 10.11 could test if the effect selector drop-down looks like this, i.e. checkmarks drawn properly, no offset/cropped list items:
image

**Comment by ronso0** [comment]
@esbrandt Could you please test Deere again? The closed effect selector should now have a dark bg color.

File: res/skins/Deere/style-mac.qss [hunk 1]
```diff
@@ -7,7 +7,7 @@
 }
 
 
-/* hack to hide the checkmark, otherwise text gets cut off
+/* Previously necessayr hack to hide the checkmark, otherwise text gets cut off
```
Comment (Be-ing) [inline res/skins/Deere/style-mac.qss]: Why not just delete the code?

**Comment by Be-ing** [comment]
IMO the contrast for highlighted menu items in Deere should still be improved.

File: res/skins/Deere/style-mac.qss (hunk 1 above)
Comment (ronso0) [reply res/skins/Deere/style-mac.qss]: For documentation. It took a while to notice and figure out the fix, so its valuble info IMO

**Comment by ronso0** [comment]
I'll take a look.

File: res/skins/Deere/style-mac.qss (hunk 1 above)
Comment (Be-ing) [reply res/skins/Deere/style-mac.qss]: I'd rather have the code deleted if it is no longer necessary. It is still in the Git history if that info is needed later.

**Comment by Holzhaus** [comment]
Thanks @ronso0, nice work!

I notices a few more issues while testing:

The context menus look bit too large on my 4K screen:

2019-11-21-142430_609x451_scrot

2019-11-21-142039_1439x1821_scrot

Same for tooltips:
2019-11-21-142631_3840x2160_scrot

The file picker (for selecting a new cover) looks broken:
2019-11-21-142105_1518x1390_scrot

**Comment by ronso0** [comment]
Thanks for testing!

Idk how the file picker is affected by the changes. 
On which OS are you?
I'll try to limit the `QScrollBar` style to `WLibrary QScrollBar, WLibrarySidebar QScrollBar`, and find a similiar selector for the dialog bg.

About the menu/tooltip font size:
we have several font issues in master, and it would be nice to solve those for 2.3:
* search box / table header don't scale with the Library font size set in Preferences
* library icons don't scale consistently with the Library font size ...
  (BPM lock, 'played' checkbox, Preview button)

For now I'll remove the fixed fon sizes from the menu, tooltip and other elements, hoping they use the OS default size.

**Comment by daschuer** [comment]
Out of curious, what are the native dimensions of the cover? 
What are your display scaling settings?

**Comment by Holzhaus** [comment]
> Out of curious, what are the native dimensions of the cover?

Either 500x500px (embedded in the FLAC tag) or 550x550px (`cover.jpg` in the same folder). I don't know which one Mixxx uses.

> What are your display scaling settings?

None AFAIC. I don't have a DE installed (using `i3wm` via `startx`) and didn't set any scaling settings. Qt seems to determine the scale factor automatically from the screen size. I can disable it via `QT_AUTO_SCREEN_SCALE_FACTOR=0`, but that's about it.

**Comment by ronso0** [comment]
> I'll try to limit the `QScrollBar` style to `WLibrary QScrollBar, WLibrarySidebar QScrollBar`

...and do the same for QHeaderView. Uff

**Comment by ywwg** [comment]
Does this PR also cover style of dialog boxes spawned by right click menus?  Right now they are hard to read on LateNight: 

image

**Comment by Holzhaus** [comment]
> @ronso0: Idk how the file picker is affected by the changes.
> On which OS are you?

I use Arch btw.

**Comment by ronso0** [comment]
> Does this PR also cover style of dialog boxes spawned by right click menus? Right now they are hard to read on LateNight:
> 
> image

I remember experiments with overview labels with those colors exactly (css green & blue) but in this branch I don't use those styles.

**Comment by ronso0** [comment]
@Holzhaus Do you see the skinned file selector scrollbars also in Deere?

> > I'll try to limit the `QScrollBar` style to `WLibrary QScrollBar, WLibrarySidebar QScrollBar`
> 
> ...and do the same for QHeaderView. Uff

Scrollbars may also pop up in the effect selector if someone has a lot of external effects enabled.
Double Ufff..

**Comment by ronso0** [comment]
@Holzhaus Could you please pull and test the cover dialog again?
With the previous commit all scrollbar and bacckground styles should be appplied to Mixxx widgets only.

**Comment by Holzhaus** [comment]
@ronso0 Voilà:
2019-11-25-202420_3840x2160_scrot

**Comment by ronso0** [comment]
alright, thx!
with 13wm only, previously the Mixxx styles must have popped up all over the place. so your screens are good for working on keepin all Mixx styles, well within Mixxx..

**Comment by ronso0** [comment]
@Holzhaus 
So this would be the next test lap: please check if the context menu in the file dialog does NOT have the LateNigth style. I'm not sure if the explicite qss selectors suffice.

If that works I'm finally close to the finish line...

**Comment by Holzhaus** [comment]
> @Holzhaus
> So this would be the next test lap: please check if the context menu in the file dialog does NOT have the LateNigth style. I'm not sure if the explicite qss selectors suffice.
> 
> If that works I'm finally close to the finish line...

Great, it works on my crappy laptop (also using i3wm):

2019-11-28-01-55-32_1366x768

I can also retest on my desktop PC tomorrow.

**Comment by Holzhaus** [comment]
However, the coverart context menu is not styled properly:

2019-11-28-020322_1366x768_scrot

**Comment by ronso0** [comment]
> However, the coverart context menu is not styled properly

Aahh, it works for WCoverArt but I have to include WSpinny QMenu as well.
Done.

The context menu in file dialogs is okay now?

**Comment by Holzhaus** [comment]
Ok, looks good to me (on LateNight at least). Two minor things I found:

The cover art in the library is not centered (don't know if this is due to qss changes):
2019-11-28-200043_914x1007_scrot

It might make sense to color the background of the fullsize coverart widget too. You can see it when resizing the window:
2019-11-28-200107_1498x1096_scrot

**Comment by ronso0** [comment]
Thanks! I'm glad it works well, so I can move the LateNight styles over to the other skins.

I didn't change any qss styles that might cause this

> The cover art in the library is not centered (don't know if this is due to qss changes):

or this 

> It might make sense to color the background of the fullsize coverart widget too. You can see it when resizing the window:

and both bugs were not there earlier. They're probably a side effect of #2247

**Comment by ronso0** [comment]
> It might make sense to color the background of the fullsize coverart widget too. You can see it when resizing the window:

Okay, you mean when manually dragging a window border? For this case it would make sense..
Didn't notice that earlier because I always resize by scrolling inside the cover window. This way it keeps its aspect ratio and there's no margin at the sides.

**Comment by Holzhaus** [comment]
Yes, I'm resizing the window manually. I'm not sure, maybe we can disable resizing the window or force the window to keep the aspect ratio while resizing.

**Comment by Holzhaus** [comment]
In any case I think these are minor problems that can be dealt with in another PR and aren't blocking the merge of this one.

**Comment by ronso0** [comment]
> Yes, I'm resizing the window manually. I'm not sure, maybe we can disable resizing the window or force the window to keep the aspect ratio while resizing.

Good idea!
When an edge is dragged try to move the edges of the other dimension so that the aspect ratio is kept. Seems overcomplicated though.
It's somehow reasonable to drag the corners of the coevr window (keep aspect ratio), but I don't see the benefit of resizing the window by dragging the edges when the actual cover view doesn't change. Also, dragging a vertical window edge, then a horizontal edge seems like a waste of time when virtually any device Mixxx runs on has scrolling abilities. for the rare case a device can't scroll (mouse wit no scroll wheel?) we could map the +/- keys to do it.

**Comment by ronso0** [comment]
@Holzhaus I'm about to include styles for `WCueMenuPopup` like for QMenu. bg color is applied, however the borders aren't drawn.
Could `Qt::WA_StyledBackground` a solution without screwing up the default Fusion theme? https://stackoverflow.com/q/52426751

I could as well exclude `WCueMenuPopup` here and just set individual Delete icons for now so we can merge this asap.

**Comment by Holzhaus** [comment]
> @ronso0: I'm about to include styles for `WCueMenuPopup` like for QMenu. bg color is applied, however the borders aren't drawn.

The button borders? Or the borders of the `WCueMenuPopup` widget itself?

> @ronso0: Could Qt::WA_StyledBackground a solution without screwing up the default Fusion theme? https://stackoverflow.com/q/52426751

I have no idea? Does styling work when you set it? I can try it tomorrow.

**Comment by ronso0** [comment]
There's no issue with the buttons, I mean the main widget border which I didn't try to style earlier.

**Comment by Holzhaus** [comment]
> > @ronso0: Could Qt::WA_StyledBackground a solution without screwing up the default Fusion theme? https://stackoverflow.com/q/52426751
> 
> I have no idea? Does styling work when you set it? I can try it tomorrow.

Yup, that works.

When I add this diff:
```diff
diff --git a/src/widget/wcuemenupopup.cpp b/src/widget/wcuemenupopup.cpp
index 215c6f4fd6..38a0542717 100644
--- a/src/widget/wcuemenupopup.cpp
+++ b/src/widget/wcuemenupopup.cpp
@@ -11,6 +11,7 @@ WCueMenuPopup::WCueMenuPopup(QWidget* parent)
         : QWidget(parent) {
     QWidget::hide();
     setWindowFlags(Qt::Popup);
+    setAttribute(Qt::WA_StyledBackground);
     setObjectName("WCueMenuPopup");

     m_pCueNumber = new QLabel(this);
```

2019-12-06-131324_712x405_scrot

Now it's also possible to add some styles:
```diff
diff --git a/res/skins/LateNight/style.qss b/res/skins/LateNight/style.qss
index 9dfbf52511..7f49a66950 100644
--- a/res/skins/LateNight/style.qss
+++ b/res/skins/LateNight/style.qss
@@ -572,6 +572,11 @@ WTime {
     font-size: 10px/10px;
   }

+WCueMenuPopup {
+    background-color: #FFFF00;
+    border: 4px solid #0000FF;
+}
+
 /********************** Decks **********************/
 #DeckRow_1_KeyVinylFx {
   qproperty-layoutAlignment: 'AlignLeft | AlignTop';
```

2019-12-06-131215_761x394_scrot

**Comment by ronso0** [comment]
> Now it's also possible to add some styles

great, thank you!
should I merge this line here or do you want to do that in a separate PR, so this here contains styles only?

**Comment by Holzhaus** [comment]
> should I merge this line here or do you want to do that in a separate PR, so this here contains styles only?

Since it's a one-line change you can apply the change here, but if you want me to open a separate PR I can do that as well.

**Comment by ronso0** [comment]
Alright, I'll add it here.

**Comment by ronso0** [comment]
and now comes the fun part: reolving conflicts with new LateNight. yeiij

**Comment by ronso0** [comment]
okay, looks like I caught all conflicts.

LateNight Claasic scheme can be tested again.
There, the cue popup is supposed to looks like this:
image

**Comment by ronso0** [comment]
As you may notice, I tweaked the deafult Delete icon a bit, namely I increased the icon size, made it brighter to fit LateNight and removed the arrow pointing into the trash bin as this was not really recognizable on my screen (15" 1366x768).
If you go along with the changes I'll use it for the other skins, too, just adapt the color to match each scheme.

**Comment by ronso0** [comment]
@Holzhaus 
nice, the qss check pointed me to a copy/paste whoopsie in Shade/style.qss: QScrollBar > Scrollbar

**Comment by ronso0** [comment]
Alrighty, this is ready to test again.

Later on I'll post a list with some aspects that should be checked, as well as some screenshots showing how the skins should look.

**Comment by Holzhaus** [comment]
What's the status of this? Is this still WIP or is it ready to test?

**Comment by ronso0** [comment]
I got some feedback on Windows from @JosepMaJAZ and I'll copy the respective changes over to the other skins.
I'm testing with Linux, and your feedback helped me discover and wipe out most issues, so I think this is ready on Linux, too.
I could still use some MacOS feedback @esbrandt ?
Screenshots will be helpful:
* library editlines
* open cue editor
* context menus
* open effect selector / fade mode selector

> Later on I'll post a list with some aspects that should be checked, as well as some screenshots showing how the skins should look.

I'll find time for this on the weekend, so everyone who's testing can compare.
If there are no major issues on all OS, especially with the menus and the hotcue editor, this is ready for merge soonish.

**Comment by Holzhaus** [comment]
Hmm, the selected color in the cue menu popup is smaller than the other ones:
2020-01-09-161946_375x258_scrot
I have no idea if this is related to your changes or not though.

Everything else LGTM ;-)

**Comment by ronso0** [comment]
Okay, cool!

> Hmm, the selected color in the cue menu popup is smaller than the other ones

The only style I added that's related to the color buttons is [this](https://github.com/mixxxdj/mixxx/pull/2363/files#diff-6761a9e770568e999bffff10ea18bf5cR2173)
```
WCueMenuPopup QPushButton:focus {
  outline: none;
}
```
and it's supposed to remove the OS-applied border and tinting from focused pushbuttons the moment they're pressed.
Idk how that would make the button shrink vertically only, though I saw it earlier already when you built the cue menu nad we played aroud with 'checked' icons.
Can you check how it looks like when you remove that `WCueMenuPopup QPushButton:focus` line?

Also, from your screenshot, the scaling doesn't seem to be apllied correctly, the Delete button should actually be bigger. But I guess that's a minor issue with all qss pixel sizes.
image

**Comment by Holzhaus** [comment]
Removing `outline: none;` does not fix the issue. I guess it gets resized when the icon is added. When I add the size explicitly, it work fine:

```diff
diff --git a/res/skins/default.qss b/res/skins/default.qss
index 667f25e100..7af74564b9 100644
--- a/res/skins/default.qss
+++ b/res/skins/default.qss
@@ -2,6 +2,11 @@ WCueMenuPopup #CueDeleteButton {
   qproperty-icon: url(:/images/ic_delete.svg);
 }

+WColorPicker QPushButton {
+  width: 50px;
+  height: 22px;
+}
+
 WColorPicker QPushButton[checked="false"] {
   qproperty-icon: none;
 }
```

**Comment by ronso0** [comment]
I fear (and already saw it soemwhere IIRC, not sure when/where) that the cue name edit line was stretched differently in skins and the color button grid stretched accordingly to fill the space underneath.

That said, I could add that fixed size to default.qss but I'm afraid to screw up more than I fix, especially considering how slow the testing/feedback/fix loop is with this PR for Win & MacOS I'm a bit reluctant to add this bit here, but rather put it on a FixLater list for after merge when this changes get much more testing due to being avalibale in builds on downloads.mixxx.org/

**Comment by ronso0** [comment]
I'm still up to post screenshots how the menus, tables & the cue popup are supposed to look like!

To anyone testing this beforehand: please post screens of some common situations like open menu etc.

**Comment by ronso0** [comment]
Here are the screenshots for comparison.
In case your view differs, please post screens.

menus-popups-etc_deere
menus-popups-etc_latenight
menus-popups-etc_shade-dark
menus-popups-etc_tango

**Comment by Holzhaus** [comment]
The text in the cue popup is unreadable in Tango and Deere.

2020-01-21-143307_3840x2160_scrot
2020-01-21-143058_3840x2160_scrot

Also, the library/context menu are quite large compared to the rest of the application, but that's probably unrelated to this branch:
2020-01-21-142940_3840x2160_scrot

**Comment by ronso0** [comment]
Thanks for testing this again & again!
Ok, so in your case `CueNumberLabel` & `CuePositionLabel` don't inherit styles from WCueMenuPopup but have to be addressed explicitely. Will fix.

I chose a rather large spacing for menu items for better readability, font size is still set by OS, which is probably also the reason for large appearance in your screenshot.
I'm curious how it looks like when you run it with scale_factor=2.0

**Comment by ronso0** [comment]
If the menu item spacing is considered to wide with 100% scaling by other users, as well, I'd decrease that, of course.

**Comment by ronso0** [comment]
remaining ToDos:
* in-depth review on macOS
* opinions on the menu items spacing

Apart from that this is ready to merge IMO so it receives broader testing from those who rely on pre-built binaries. I bet we'll discover minor issues anyway, but those can be fixed later on.

**Comment by Holzhaus** [comment]
When using Tango, the content of the label field is not visible unless selected:

2020-01-23-132700_619x358_scrot

> opinions on the menu items spacing

I think the spacing is ok. For some reason, I feel like all custom UI elements (i.e. everything execept library and menu bar) are too small on my 4K screen (or library and menu bar are huge when using `QT_SCALE_FACTOR=2`), but that issue is unrealated to this PR.

**Comment by ronso0** [comment]
> When using Tango, the content of the label field is not visible unless selected:

Fixed. I just double-checked the other QLineEdit styles and I think this was the last missing piece in this regard.

> library and menu bar are huge when using QT_SCALE_FACTOR=2

you refer to the table header?

**Comment by Holzhaus** [comment]
Nope, the top menu bar. The table header is actually fine, it's just that the entries are large in comparison:
2020-01-23-153221_3840x2160_scrot

2020-01-23-153024_3840x2160_scrot

**Comment by Holzhaus** [comment]
By the way, in Tango there are some icons visible below the "No Headphones configured". I think this is a bug, but this also happens in master:
2020-01-23-153749_158x47_scrot

**Comment by ronso0** [comment]
> By the way, in Tango there are some icons visible below the "No Headphones configured".

It's opaque now.
The PreviewDeck cover I'd leave translucent, though, so there's some feedback for when loading / previewing a track from the library. It'll be hard to make the (UX) situations with unconfigured devices consistent for mic, vinyl and headphone (how control canges are processed / allowed)...

**Comment by Holzhaus** [comment]
This is ready to merge, but needs some MacOS testing, right?

**Comment by Holzhaus** [comment]
I just checked again. In direct comparison, the spacing between the items could be a bit smaller. But I guess that's just personal preference.


2020-02-07-133748_602x461_scrot
2020-02-07-134128_466x422_scrot

**Comment by ronso0** [comment]
> This is ready to merge, but needs some MacOS testing, right?

Yep, final feedback from MacOS is missing.
Besides the merge conflicts, this is left to do:
* somehow remove middle elide of effect names in LateNight, this is impractical n small screens (can't tell what's causing this)
* reduce menu item spacing

**Comment by ronso0** [comment]
> somehow remove middle elide of effect names in LateNight

for some reason the label of the (closed) effect selector is shown elided sometimes, indeveloper mode though. strange..

**Comment by ronso0** [comment]
@Holzhaus the qss check has proven to be very helpful once more. Thx!

**Comment by Holzhaus** [comment]
Is this ready to review and merge?

**Comment by ronso0** [comment]
this is ready for some time already.
still needs testing/confirmation on MacOS & WIN.
can't tell who to ping for that right now.

I still think we can merge and fix minor issues afterwards when nightly builds are avaliable dor everyone

**Comment by Holzhaus** [comment]
We could merge this now and get feedback during the 2.3 beta phase.

**Comment by ronso0** [comment]
Yup! +1

**Comment by Holzhaus** [comment]
Done. Thank you for your work!

**Pull request merged**